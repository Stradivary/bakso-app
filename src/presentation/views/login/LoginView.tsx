import {
  Box,
  Button,
  Checkbox,
  Container,
  Flex,
  Image,
  Select,
  Stack,
  Text,
  TextInput,
  Title,
} from "@mantine/core";
import { Turnstile, TurnstileInstance } from "@marsidev/react-turnstile";
import { useRef, useState } from "react";
import { Controller } from "react-hook-form";
import { useLoginViewModel } from "./useLoginViewModel";

export function LoginView() {
  const {
    control,
    isValid,
    handleSubmit,
    handleAuth,
    errors,
    isSubmitting,
    canJoin,
    setValue,
    handleOnChange,
  } = useLoginViewModel();

  const turnstileRef = useRef<TurnstileInstance>(null);

  const [token, setToken] = useState<string>();

  return (
    <Flex h="100dvh" align="center" bg="#EFF1F4">
      <Container size="xs" mx="auto" ta="center">
        <Box mb="md">
          <Image w={150} mx="auto" src="/video.png" alt="Illustration" />
        </Box>
        <Title order={2}>Verifikasi</Title>
        <Text mt="sm" mb="lg">
          Masukkan nama dan role Anda di bawah ini:
        </Text>

        <form onSubmit={handleSubmit(handleAuth)}>
          <Stack gap={16} ta="start">
            <input type="hidden" name="token" value={token} />
            <Controller
              name="name"
              control={control}
              render={({ field }) => (
                <TextInput
                  label="Nama"
                  placeholder="Masukkan nama"
                  {...field}
                  error={errors.name}
                />
              )}
            />

            <Controller
              name="role"
              control={control}
              render={({ field }) => (
                <Select
                  label="Role"
                  placeholder="Masukkan role"
                  rightSection={<></>}
                  data={[
                    { value: "buyer", label: "Pelanggan" },
                    { value: "seller", label: "Tukang Bakso" },
                  ]}
                  {...field}
                  error={errors.role}
                />
              )}
            />

            <Controller
              name="agreed"
              control={control}
              render={({ field: { value, onChange, ...fld } }) => (
                <Checkbox
                  label="Dengan menggunakan aplikasi ini Anda telah setuju untuk membagikan lokasi Anda kepada para tukang Bakso Keliling."
                  checked={value}
                  color="secondary.9"
                  onChange={(e) => handleOnChange(onChange, e)}
                  {...fld}
                  error={errors.agreed}
                />
              )}
            />
            <Turnstile
              ref={turnstileRef}
              siteKey="0x4AAAAAAA0a2jjKL7pnbfjV"
              options={{
                action: "submit-form",
                theme: "light",
                size: "flexible",

                language: "id",
              }}
              scriptOptions={{
                appendTo: "body",
                defer: true,
              }}
              onSuccess={(token) => {
                setToken(token);
                setValue("verified", true);
              }}
            />

            <Button
              fullWidth
              disabled={canJoin && isValid}
              color="red"
              size="md"
              radius={56}
              type="submit"
              loading={isSubmitting}
            >
              Join
            </Button>
          </Stack>
        </form>
      </Container>
    </Flex>
  );
}

export default LoginView;
