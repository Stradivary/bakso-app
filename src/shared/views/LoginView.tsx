import {
  Box,
  Button,
  Checkbox,
  Container,
  Flex,
  Image,
  Select,
  Stack,
  Text,
  TextInput,
  Title,
} from "@mantine/core";
import { Controller } from "react-hook-form";
import { useLoginViewModel } from "../viewModels/useLoginViewModel";

export function LoginView() {
  const {
    control,
    handleSubmit,
    handleAuth,
    errors,
    isSubmitting,
    canJoin,
    handleOnChange,
  } = useLoginViewModel();

  return (
    <Flex h="100dvh" align="center" bg="#EFF1F4">
      <Container size="xs" mx="auto" ta="center">
        <Box mb="md">
          <Image w={150} mx="auto" src="/video.png" alt="Illustration" />
        </Box>
        <Title order={2}>Verifikasi</Title>
        <Text mt="sm" mb="lg">
          Masukkan nama dan role Anda di bawah ini:
        </Text>

        <form onSubmit={handleSubmit(handleAuth)}>
          <Stack gap={16} ta="start">
            <Controller
              name="name"
              control={control}
              rules={{
                required: "Nama anda dibutuhkan",
                maxLength: { value: 60, message: "Nama anda terlalu panjang" },
              }}
              render={({ field }) => (
                <TextInput
                  label="Nama"
                  placeholder="Masukkan nama"
                  {...field}
                  error={errors.name}
                />
              )}
            />

            <Controller
              name="role"
              control={control}
              rules={{
                required:
                  "Anda harus memilih diantara customer atau Tukang Bakso",
              }}
              render={({ field }) => (
                <Select
                  label="Role"
                  placeholder="Masukkan role"
                  rightSection={<></>}
                  data={[
                    { value: "buyer", label: "Pelanggan" },
                    { value: "seller", label: "Tukang Bakso" },
                  ]}
                  {...field}
                  error={errors.role}
                />
              )}
            />

            <Controller
              name="agreed"
              control={control}
              rules={{ required: "Anda harus menyetujui pernyataan di atas" }}
              render={({ field: { value, onChange, ...fld } }) => (
                <Checkbox
                  label="Dengan menggunakan aplikasi ini Anda telah setuju untuk membagikan lokasi Anda kepada para tukang Bakso Keliling."
                  checked={value}
                  color="secondary.9"
                  onChange={(e) => handleOnChange(onChange, e)}
                  {...fld}
                  error={errors.agreed}
                />
              )}
            />

            <Button
              fullWidth
              disabled={canJoin}
              color="red"
              size="md"
              radius={56}
              type="submit"
              loading={isSubmitting}
            >
              Join
            </Button>
          </Stack>
        </form>
      </Container>
    </Flex>
  );
}

export default LoginView;
