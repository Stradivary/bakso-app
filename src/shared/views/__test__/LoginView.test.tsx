import { MantineProvider } from "@mantine/core";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { beforeEach, describe, expect, it, vi } from "vitest";
import { LoginView } from "../LoginView";

vi.mock("react-router-dom", async (importOriginal) => {
  const actual = (await importOriginal()) as Record<string, never>;
  return {
    ...actual,
    useNavigate: vi.fn().mockImplementation((a) => console.log(a)),
  };
});
vi.mock("../../hooks/useAuth", () => ({
  useAuth: vi.fn().mockReturnValue({
    login: vi.fn(),
  }),
}));

vi.mock("../../hooks/useLocation", () => ({
  useLocation: vi.fn(),
}));

vi.mock("../../hooks/useLocationUpdater", () => ({
  useLocationUpdater: vi.fn(),
}));

vi.mock("../../hooks/useRequestLocation", () => ({
  useRequestLocation: vi.fn().mockReturnValue({
    latitude: 0,
    longitude: 0,
    error: "",
    isLoading: false,
    requestLocation: vi.fn(),
  }),
}));

describe("LoginView", () => {
  beforeEach(() => {
    // Reset mocks
    vi.clearAllMocks();
  });

  it("renders login form correctly", () => {
    const { container } = render(
      <MantineProvider>
        <MemoryRouter>
          <LoginView />
        </MemoryRouter>
      </MantineProvider>,
    );

    expect(container).toMatchSnapshot();
  });
});
