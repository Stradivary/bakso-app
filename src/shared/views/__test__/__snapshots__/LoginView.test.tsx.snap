// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginView > renders login form correctly 1`] = `
<div>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <div
    class="m_8bffd616 mantine-Flex-root __m__-r0"
    style="align-items: center; background: rgb(239, 241, 244);"
  >
    <div
      class="m_7485cace mantine-Container-root"
      data-size="xs"
      style="--container-size: var(--container-size-xs); margin-inline: auto; text-align: center;"
    >
      <div
        class=""
      >
        <img
          alt="Illustration"
          class="m_9e117634 mantine-Image-root"
          src="/video.png"
          style="margin-inline: auto;"
        />
      </div>
      <h2
        class="m_8a5d1357 mantine-Title-root"
        data-order="2"
        style="--title-fw: var(--mantine-h2-font-weight); --title-lh: var(--mantine-h2-line-height); --title-fz: var(--mantine-h2-font-size);"
      >
        Verifikasi
      </h2>
      <p
        class="mantine-focus-auto m_b6d8b162 mantine-Text-root"
      >
        Masukkan nama dan role Anda di bawah ini:
      </p>
      <form>
        <div
          class="m_6d731127 mantine-Stack-root"
          style="--stack-gap: calc(1rem * var(--mantine-scale)); --stack-align: stretch; --stack-justify: flex-start; text-align: start;"
        >
          <div
            class="m_46b77525 mantine-InputWrapper-root mantine-TextInput-root"
          >
            <label
              class="m_8fdc1311 mantine-InputWrapper-label mantine-TextInput-label"
              for=""
              id=""
            >
              Nama
            </label>
            <div
              class="m_6c018570 mantine-Input-wrapper mantine-TextInput-wrapper"
              data-variant="default"
            >
              <input
                aria-invalid="false"
                class="m_8fb7ebe7 mantine-Input-input mantine-TextInput-input"
                data-variant="default"
                id=""
                name="name"
                placeholder="Masukkan nama"
                value=""
              />
            </div>
          </div>
          <div
            class="m_46b77525 mantine-InputWrapper-root mantine-Select-root"
          >
            <label
              class="m_8fdc1311 mantine-InputWrapper-label mantine-Select-label"
              for=""
              id=""
            >
              Role
            </label>
            <div
              class="m_6c018570 mantine-Input-wrapper mantine-Select-wrapper"
              data-pointer="true"
              data-variant="default"
              data-with-right-section="true"
              style="--input-right-section-pointer-events: none;"
            >
              <input
                aria-haspopup="listbox"
                aria-invalid="false"
                autocomplete="off"
                class="m_8fb7ebe7 mantine-Input-input mantine-Select-input"
                data-variant="default"
                id=""
                placeholder="Masukkan role"
                readonly=""
                value=""
              />
              <div
                class="m_82577fc2 mantine-Input-section mantine-Select-section"
                data-position="right"
              />
            </div>
          </div>
          <input
            name="role"
            type="hidden"
            value=""
          />
          <div
            class="m_bf2d988c mantine-Checkbox-root m_5f75b09e mantine-Checkbox-root"
            data-label-position="right"
            style="--checkbox-color: secondary.9;"
          >
            <div
              class="m_5f6e695e mantine-Checkbox-body"
            >
              <div
                class="m_26062bec mantine-Checkbox-inner"
                data-label-position="right"
              >
                <input
                  class="mantine-focus-auto m_26063560 mantine-Checkbox-input"
                  id=""
                  name="agreed"
                  type="checkbox"
                />
                <svg
                  aria-hidden="true"
                  class="m_bf295423 mantine-Checkbox-icon"
                  fill="none"
                  viewBox="0 0 10 7"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M4 4.586L1.707 2.293A1 1 0 1 0 .293 3.707l3 3a.997.997 0 0 0 1.414 0l5-5A1 1 0 1 0 8.293.293L4 4.586z"
                    fill="currentColor"
                    fill-rule="evenodd"
                  />
                </svg>
              </div>
              <div
                class="m_d3ea56bb mantine-Checkbox-labelWrapper"
              >
                <label
                  class="m_8ee546b8 mantine-Checkbox-label"
                  for=""
                >
                  Dengan menggunakan aplikasi ini Anda telah setuju untuk membagikan lokasi Anda kepada para tukang Bakso Keliling.
                </label>
              </div>
            </div>
          </div>
          <button
            class="mantine-focus-auto m_77c9d27d mantine-Button-root m_87cf2631 mantine-UnstyledButton-root"
            data-block="true"
            data-disabled="true"
            data-size="md"
            disabled=""
            style="--button-height: var(--button-height-md); --button-padding-x: var(--button-padding-x-md); --button-fz: var(--mantine-font-size-md); --button-radius: calc(3.5rem * var(--mantine-scale)); --button-bg: var(--mantine-color-red-filled); --button-hover: var(--mantine-color-red-filled-hover); --button-color: var(--mantine-color-white); --button-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent;"
            type="submit"
          >
            <span
              class="m_80f1301b mantine-Button-inner"
            >
              <span
                class="m_811560b9 mantine-Button-label"
              >
                Join
              </span>
            </span>
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
